node/core/pvf/src/prepare/worker.rs

```diff
@@ -175,6 +181,15 @@ async fn handle_response(
 		Ok(result) => result,
 		// Timed out on the child. This should already be logged by the child.
 		Err(PrepareError::TimedOut) => return Outcome::TimedOut,
+		// Worker reported version mismatch
+		Err(PrepareError::VersionMismatch) => {
+			gum::error!(
+				target: LOG_TARGET,
+				%worker_pid,
+				"node and worker version mismatch",
+			);
+			std::process::exit(1);
```

Comment:
I think you can send a signal to the overseer to let it shutdown?

Reply:
It's PVF host, it isn't a subsystem by itself, and it's not aware of overseer's existence :roll_eyes: 
To handle it the right way, the error should be propagated all the way up to the candidate validation and PVF pre-check subsystems and handled by them separately. Do you think it's worth the effort?

Reply:
You shouldn't just kill the process here, you should bubble up this error and let the host kill the worker properly. Look at how the host handles e.g. `TimedOut` for an example. Also, I'm pretty sure this code is running on the host so `exit` would kill the node, not the worker, which I think is not what we want? (Haven't followed the disputes issue that closely over the weekend.)

Reply:
> Also, I'm pretty sure this code is running on the host so `exit` would kill the node, not the worker

Yes, it's exactly what we want. At this point, we found out that the node owner screwed up the node upgrade process and we're still running an old node software, but the binary is already new, and we're trying to execute a new worker from the old node. In that case, we want to tear the whole node down and let the node owner handle its restart.

Reply:
Oh, got it. Is any clean-up of the workers necessary, or do they get killed automatically as child processes?

Reply:
Children are not killed automatically, but if the worker recognizes `version_mismatch`, it will `return` from the main loop which will terminate the process.

Other workers may still be alive at the moment, but after they get some outcome from their jobs they'll try to send it to the node through the unix socket that is already closed, which will also force them to terminate.

We can terminate them manually (dropping the workers' handles) before tearing down the node, but why? They'll exit anyway.

Reply:
Please document this all clearly above the `exit`. There are already a number of silent assumptions being made throughout the pvf code, and we should at least have comments to make things more explicit. It is also good just to draw attention to the `exit` because it is quite an extreme measure to suddenly shut down the entire node.

Reply:
We need to wait for the workers to shutdown. Otherwise we may put stuff into the cache that is invalid. Let's assume the following. Node at version X is spawning a preparation worker. Node at version X is restarted and starts again as node at version x + 1. The worker finishes and writes the wasm artifact into the cache. Then the cache contains an artifact from version X.

Reply:
Good point. Unfortunately this means potentially waiting the full lenient timeout duration for a worker to exit. I think it would make sense to just kill them forcefully so we don't wait.

Alternatively, we could detect which artifacts were from an old version. This is useful anyways if we ever stop deleting the cache on restart (there is an ongoing discussion about this). But if we knew that artifacts were from an old worker and invalid, then we wouldn't need to shutdown workers.

Reply:
It makes sense to implement @koute suggestion as well and include polkadot's version into the artifact path. This way old version can't override an artifact for the new node.

Reply:
> Good point. Unfortunately this means potentially waiting the full lenient timeout duration for a worker to exit. I think it would make sense to just kill them forcefully so we don't wait.

Yes for sure we should kill the workers, they don't do anything important that could get lost when we kill them.

> It makes sense to implement @koute suggestion as well and include polkadot's version into the artifact path. This way old version can't override an artifact for the new node.

This should be a separate pr.

Reply:
> We need to wait for the workers to shutdown.

Good point! But I've reviewed the code and looks like @pepyakin thought this through before us.

`start_work()` on the host side provides the worker with a temporary path where it should store the artifact. If the worker reports a positive result, then in `handle_result()`, again on the host side, the artifact gets renamed to its real `artifact_id` path. If the host dies between `start_work()` and `handle_result()`, the only leftover is a stale tmp file which will be pruned on the next node restart.

It's still a good idea to kill preparation workers on version mismatch.

Reply:
@koute you mean some random number we generate on node startup and use throughout the node instance lifetime? I think it's a good approach also, but your first idea about including the node version number to the artifact name is somewhat more... Deterministic, I'd say? :)